openwisp_utils/admin_theme/schema.py

```diff
@@ -0,0 +1,48 @@
+from django.core.exceptions import ImproperlyConfigured
+
+from ..utils import SortedOrderedDict
+
+DASHBOARD_SCHEMA = SortedOrderedDict()
+
+
+def _validate_element_params(schema):
+    query_params = schema.get('query_params', None)
+
+    assert query_params is not None
+    assert 'app_label' in query_params
+    assert 'model' in query_params
+    assert 'group_by' in query_params
+    try:
+        assert 'order' in schema
+    except AssertionError:
+        schema['order'] = len(DASHBOARD_SCHEMA.keys())
+    return schema
+
+
+def register_dashboard_element(element_name, element_params):
+    """
+    Registers a dashboard element
+    register_dashboard_element(str, dict)
+    """
+    if not isinstance(element_name, str):
+        raise ImproperlyConfigured('Dashboard element name should be type `str`.')
+    if not isinstance(element_params, dict):
+        raise ImproperlyConfigured(
+            'Dashboard element parameters should be type `dict`.'
+        )
+    if element_name in DASHBOARD_SCHEMA:
+        raise ImproperlyConfigured(
+            f'{element_name} is an already registered Dashboard Element.'
+        )
+
+    validated_element_params = _validate_element_params(element_params)
+    DASHBOARD_SCHEMA.update({element_name: validated_element_params})
```

Comment:
An alternative could be to use the order as the dict key (since the order is more important than the name in this case).
This would allow to avoid having to sort it at each registration, we can just leave it unsorted and then sort it in `OpenwispAdminSite` before it will be used.

```python
register_dashboard_element(
    position=1
    params={
        'name': 'Operator Project Distribution',
        'query_params': {
            'app_label': 'test_project',
            'model': 'operator',
            'group_by': 'project__name',
        },
        'colors': {'Utils': 'red', 'User': 'orange'},
    },
)
```

```python
def dashboard(self, request, extra_context=None):
    # sort a copy of DASHBOARD_SCHEMA here based on dict keys
    ...
```

Reply:
> This would allow to avoid having to sort it at each registration, we can just leave it unsorted and then sort it in OpenwispAdminSite before it will be used 

I thought about this but realised that it will introduce a delay while loading the page. All the elements are supposed to be registered through `ready` method of the application. Sorting while adding a new element will only add delay on the startup of the project. After that, there will be no operation needed for sorting the `DASHBOARD_CONFIGURATION`. 

> ```python
> register_dashboard_element(
>     position=1
>     params={
>         'name': 'Operator Project Distribution',
>         'query_params': {
>             'app_label': 'test_project',
>             'model': 'operator',
>             'group_by': 'project__name',
>         },
>         'colors': {'Utils': 'red', 'User': 'orange'},
>     },
> )
> ```

I like using position as key in the dictionary, just like a [TreeMap](https://www.geeksforgeeks.org/treemap-in-java/)

Reply:
I would not worry about this delay @pandafy, since it's negligible, the real delay is caused by the DB queries, that's the part which will most probably hit us with pain in the near future and we'll need to cache, but we'll get back to this issue later on.

If you want to sort each time anyway during registration it's fine as well for me.